libbeat/autodiscover/providers/kubernetes/kubernetes.go

```diff
@@ -279,7 +279,9 @@ func NewLeaderElectionManager(
 		Name:      cfg.LeaderLease,
 		Namespace: ns,
 	}
-	metaUID := lease.GetObjectMeta().GetUID()
+
+	var eventID string
```

Comment:
Is there any possibility of data races with this? Presumably OnStoppedLeading and OnStartedLeading should never be called concurrently...

Reply:
Yes, you are correct. The code for running the leader election is [in these lines](https://github.com/kubernetes/client-go/blob/1518fca9f06c6a73fc091535b8966c71704e657b/tools/leaderelection/leaderelection.go#L204-L214), and `startLeading` is running as a go routine.  It also says:

> // LeaderCallbacks are callbacks that are triggered during certain
// lifecycle events of the LeaderElector. These are invoked asynchronously.

So I will check for racing conditions and try to prevent it from happening.

Reply:
I tried to do a unit test to check the racing condition:

1. I could never find it, I think that is weird. Could it be possible that it never happens?
2. To cause a lease renewal, I would have to add a timer to ensure a new instance has the lease. I believe a good time to wait could be `(leaseDuration + retryPeriod) * 2`, but this means our test will wait 34s at most for one iteration. I was testing with multiple, so it could take 2 minutes. Is it good idea to make a commit with such a long test since it would be triggered in other PRs? What do you think?

Reply:
> To cause a lease renewal, I would have to add a timer to ensure a new instance has the lease. I believe a good time to wait could be (leaseDuration + retryPeriod) * 2, but this means our test will wait 34s at most for one iteration. I was testing with multiple, so it could take 2 minutes. Is it good idea to make a commit with such a long test since it would be triggered in other PRs? What do you think?

I agree, we should avoid such long running tests. We could perhaps make the lease duration configurable. Another option would be to pass in a fake clock, like https://github.com/jonboulle/clockwork, and then we can control it in the test.

Reply:
I added a unit test. It still takes a few seconds to complete, at most 30s. I could not reduce the lease duration fields anymore, because it was causing unexpected lease renewals (I am guessing for network issues).